File: client/src/app/classes/game-play/timer.ts
import { Speed } from '@common/enums/game-play/speed';
import { TimerConfig } from '@app/interfaces/timer/timer-config';

export class Timer {
    speed: Speed;
    timerId: number | undefined;
    private start: number;
    private remaining: number;
    private callback: (value: unknown) => void;
    private repeat: boolean;

    constructor(callback: (value: unknown) => void, config: TimerConfig) {
        this.callback = callback;
        this.remaining = config.delay;
        this.speed = config.speed;
        this.repeat = config.repeat;
        this.resume();
    }

    pause() {
        if (!this.timerId) {
            return;
        }
        window.clearTimeout(this.timerId);
        this.timerId = undefined;
        if (!this.repeat) this.remaining -= (Date.now() - this.start) * this.speed;
    }

    resume() {
        if (this.timerId) {
            return;
        }
        this.start = Date.now();
        if (this.repeat) {
            this.timerId = window.setInterval(this.callback, this.remaining / this.speed);
        } else {
            this.timerId = window.setTimeout(this.callback, this.remaining / this.speed);
        }
    }
}

File: flutter_project/lib/classes/replay/timer.dart

import 'dart:async';
import 'package:flutter_project/common/enum.dart';

class TimerConfig {
  final double delay;
  final Speed speed;
  final bool repeat;

  TimerConfig({required this.delay, required this.speed, required this.repeat});
}

class CustomTimer {
  Speed speed;
  Timer? timer;
  late int start;
  double remaining;
  final void Function() callback;
  final bool repeat;

  CustomTimer(this.callback, TimerConfig config)
      : speed = config.speed,
        remaining = config.delay,
        repeat = config.repeat {
    resume();
  }

  void pause() {
    if (timer == null) return;

    timer!.cancel();
    timer = null;
    if (!repeat) remaining -= (DateTime.now().millisecondsSinceEpoch - start) * speed.value;
  }

  void resume() {
    if (timer != null) return;

    start = DateTime.now().millisecondsSinceEpoch;
    if (repeat) {
      timer = Timer.periodic(Duration(milliseconds: remaining ~/ speed.value), (Timer t) => callback());
    } else {
      timer = Timer(Duration(milliseconds: remaining ~/ speed.value), callback);
    }
  }
}